### PR TITLE
Add docs about VolumeAttributesClass API (KEP-3751)

### PR DESCRIPTION
**Pull request by sunnylovestiramisu**
https://github.com/kubernetes/enhancements/issues/3751

https://github.com/kubernetes/kubernetes/pull/121104

Doc draft in [Google doc](https://docs.google.com/document/d/123LvscAa5KBjpycb6ajrxozxriQf8txBlgOSlErpvsk/edit)

### CODE REVIEW

**Comment by sftim** [comment]
/retitle [WIP] Add post-release blog article about  VolumeAttributesClass KEP-3751
/sig storage

/hold
If this PR is adding a blog article, it should target _main_ and be held until after the release (reviews and approval should happen in the meantime)

**Comment by sunnylovestiramisu** [comment]
@sftim The instruction link I got says create an PR against dev-1.29, it this not the case anymore? Can you send me the correct documentaion?

**Comment by sftim** [comment]
> The instruction link I got says create an PR against dev-1.29, it this not the case anymore? Can you send me the correct documentaion?

You can turn this into a documentation PR, in which case the branch is right but the path is wrong: change `content/en/docs/*` not `content/en/blog/*` (and the PR title). Then force-push.

Right now this has a mismatch because it changes a blog path. PRs that add blog articles never go on release branches.

**Comment by sunnylovestiramisu** [comment]
@sftim  which folder under docs I should post for a new feature? The structure is not clear to me.

**Comment by xing-yang** [comment]
@sunnylovestiramisu I think you just need to rename the title of this PR.  It says "Add post-release blog article about VolumeAttributesClass KEP-3751".  Just change that to "Add doc about VolumeAttributesClass KEP-3751".

**Comment by xing-yang** [comment]
It looks like "area/blog" label was added by the bot automatically probably because "blog" is in the PR title.

/area blog cancel

**Comment by sftim** [comment]
/retitle [WIP] Add docs about VolumeAttributesClass API (KEP-3751)

**Comment by sftim** [comment]
BTW our bot adds the label based on the paths in the PR, but doesn't clear the label if you force-push to remove the relevant change.

**Comment by Princesso** [comment]
Hello @sunnylovestiramisu 👋, I cannot see any file changes on this PR. Please take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before Tuesday 14th November 2023. Thank you!

**Comment by Princesso** [comment]
Hello @sunnylovestiramisu 👋, gentle reminder to take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before Tuesday 14th November 2023. Thank you!

**Comment by sunnylovestiramisu** [comment]
We are tracking the deadline, but I did not get time to work on it yet.

**Comment by Princesso** [comment]
Hello @sunnylovestiramisu 👋, gentle reminder to take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before tomorrow, Tuesday 14th November 2023. Thank you!

**Comment by sunnylovestiramisu** [comment]
@Princesso  It is ready for review :)

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (tengqm) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:1]: Each concept page has a frontmatter, separated from the main doc using `---`.
Please check other concepts page for a reference.

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (tengqm) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:3]: Please save some extra text when appropriate. Also please use relative paths for links. Please wrap the long lines for ease of PR review, change tracking and
downstream localization.

```suggestion
This page assumes that you are familiar with [StorageClasses](/docs/concepts/storage/storage-classes/),
[volumes](/docs/concepts/storage/volumes/) and [PersistentVolumes](/docs/concepts/storage/persistent-volumes/)
in Kubernetes.
```

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (tengqm) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:9]:
```suggestion
A VolumeAttributesClass provides a way for administrators to describe the mutable
"classes" of storage they offer. Different classes might map to different quality-of-service levels.
Kubernetes itself is unopinionated about what these classes represent. 
```

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (tengqm) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:33]: Where does this `VolumeAttributes` come from?
Is it always a plural?

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (carlory) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:93]: > When a parameter is omitted, some default is used.

After a user apply the PVC with a different VolumeAttributesClass, if some parameters are omitted in the target VolumeAttributesClass, the default value of the parameters may be used but it is not guaranteed. Because its behavior is determined by a specific CSI driver implementation. Please refer to the related CSI driver documentation for more details.

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (carlory) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:91]: I think it's better if there's a description about the relation and difference between the VAC's parameters and the SC's parameters. 

for SC, the parameters fields can holds both immutable volume attributes and mutable volume attributes. 
for VAC, the parameter fields should hold mutable volume attributes and takes precedence over the SC's parameters fields if there's any conflict between them when creating a volume.

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (Ritikaa96) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:37]:
```suggestion
You are not restricted to specifying the [kubernetes-csi/external-provisioner](https://github.com/kubernetes-csi/external-provisioner). You can also run and specify external provisioners,
which are independent programs that follow a specification defined by Kubernetes.
Authors of external provisioners have full discretion over where their code lives, how
the provisioner is shipped, how it needs to be run, what volume plugin it uses (including Flex), etc. 
```

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (Ritikaa96) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:93]:
```suggestion
Volume Classes have parameters that describe volumes belonging to them. Different parameters may be accepted
depending on the provisioner or the resizer. For example, the value `4000`, for the parameter `iops`,
and the parameter `throughput` are specific to GCE PD. When a parameter is omitted, the default is used.
```

**Review by Ritikaa96 — CHANGES_REQUESTED** [review]

**Comment by Princesso** [comment]
Hello @sunnylovestiramisu 👋, gentle reminder to get your PR merged before the deadline, Tuesday 28th November 2023. Thank you!

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (Ritikaa96) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:95]:
```suggestion
There can be at most 512 parameters defined for a VolumeAttributesClass.
The total length of the parameters object including its keys and values cannot exceed 256 KiB.
```

**Review by Ritikaa96 — CHANGES_REQUESTED** [review]

**Comment by katcosgrove** [comment]
Hello @sunnylovestiramisu! The deadline to have this PR merged is **Tuesday, November 28**. Enhancements that do not include required documentation may not be included in the release. Is there anything I can do to help get this ready to merge? Thanks!

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sunnylovestiramisu) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:33]: It is what we decided in the KEP design. Because these are volume attributes like iops, throughput, type etc. but they are all different depending on different vendors

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sunnylovestiramisu) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:91]: for SC, the parameters fields can holds both immutable volume attributes and mutable volume attributes. This may not be accurate, we may want to encourage users to put mutable attributes in VAC instead of SC. 

We

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (xing-yang) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:47]: VolumeAttributes -> VolumeAttributesClass?

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (xing-yang) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:54]: Can you remove " (including Flex)"?  Flex volume is already deprecated so we don't want to mention it as an example.

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (xing-yang) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:47]: The "provisioner" field you referred here is the "driverName" field?  Can you clarify?

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (xing-yang) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:59]: VolumeAttributes -> VolumeAttributesClass

The "resizer" field you referred here is the "driverName" field?  Can you clarify?

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sunnylovestiramisu) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:47]: okay let me add driverName

**Comment by Ritikaa96** [comment]
this lgtm 
/lgtm

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (Ritikaa96) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:47]:
```suggestion
Each VolumeAttributesClass has a provisioner that determines what volume plugin is used for provisioning PVs. The field `driverName` must be specified. 
```

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (Ritikaa96) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:59]:
```suggestion
Each VolumeAttributesClass has a resizer that determines what volume plugin is used for modifying PVs. The field `driverName` must be specified. 
```

Comment (Ritikaa96) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:59]: Reference to :https://kubernetes.io/docs/contribute/style/style-guide/#use-code-style-for-object-field-names-and-namespaces

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (Ritikaa96) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:30]:
```suggestion
While the name of a VolumeAttributesClass object in a `PersistentVolumeClaim` is mutable, the parameters in an existing class are immutable.
```

**Review by Ritikaa96 — CHANGES_REQUESTED** [review]

**Comment by sunnylovestiramisu** [comment]
@Ritikaa96 
@xing-yang 

Updated

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:17]: (nit) the introduction was:

> This page assumes that you are familiar with [StorageClasses](/docs/concepts/storage/storage-classes/),
> [volumes](/docs/concepts/storage/volumes/) and [PersistentVolumes](/docs/concepts/storage/persistent-volumes/)
> in Kubernetes.

```suggestion
```

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:24]:
```suggestion
## The VolumeAttributesClass API
```

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:10]:
```suggestion

{{< feature-state for_k8s_version="v1.29" state="alpha" >}}

```

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:25]: Do you need to enable any feature gates and API groups to use this API?

**Review by sftim — COMMENTED** [review]
Thanks.

- We should tell readers that this API is alpha and is disabled by default.
- We should edit the PersistentVolumes page to link to this new page.

/lgtm cancel
because of the marked-as-alpha detail

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sunnylovestiramisu) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:24]: This is in consistent with the existing Storage Class site? Do we need to change the Storage Class page as well?

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sunnylovestiramisu) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:25]: Yes feature gates, it is captured in the blog post.

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sunnylovestiramisu) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:17]: What do you mean here? Remove the introduction title completely or move it to above?

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sunnylovestiramisu) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:24]: On the storage class page we call this session ## The StorageClass Resource

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sunnylovestiramisu) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:17]: And do we change the StorageClass page as well if we change the format here?

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sftim) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:26]: It'd be nice to improve the docs about StorageClasses, but that can happen in a separate PR.

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sftim) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:27]: Cool. Please cover the details in the documentation as well.

**Comment by sunnylovestiramisu** [comment]
@sftim "We should edit the PersistentVolumes page to link to this new page." do we also add Storage Class on the PersistentVolumes page as well? Currently it is missing.

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sunnylovestiramisu) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:26]: Changed both to use ``API``

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sunnylovestiramisu) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:27]: Which session I should add this? Should it be in introduction?

**Comment by sftim** [comment]
> > We should edit the PersistentVolumes page to link to this new page.
>
> do we also add Storage Class on the PersistentVolumes page as well? Currently it is missing.

It'd be nice to improve the docs about StorageClasses, but that can happen in a separate PR.

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (reylejano) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:37]: I don't see that feature gate on the feature gates page for the `dev-1.29` branch
https://github.com/kubernetes/website/blob/dev-1.29/content/en/docs/reference/command-line-tools-reference/feature-gates.md

If it's an alpha feature gate then it should be added

**Comment by drewhagen** [comment]
The deadline to merge these documentation PRs is today.
@saad-ali as lead of sig/storage, would you be able to help us complete a final review and get this approved? Thanks!

**Comment by saad-ali** [comment]
Sure taking a look

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (saad-ali) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:85]: This should be silver right?

**Review by saad-ali — COMMENTED** [review]
Overall content is accurate so it LGTM, one small comment.

For beta we can come back and add more context and background explaining more the difference between StorageClass and VolumeAttributeClass

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sunnylovestiramisu) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:85]: Oops yes, changing it now.

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sunnylovestiramisu) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:37]: Updated the feature-gates.md

**Review by saad-ali — APPROVED** [review]
/lgtm
/approve

Thanks @sunnylovestiramisu for the hard work on this!

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sftim) [reply content/en/docs/concepts/storage/volume-attributes-classes.md:19]: I recommend removing the “Introduction” heading; we avoid having a heading with that name.

No need for _this_ PR to improve the StorageClass docs.

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:26]:
```suggestion
If you want to test the feature whilst it's alpha, you need to enable the `VolumeAttributesClass`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) for the kube-controller-manager and the kube-apiserver. You use the `--feature-gates` command line argument:
```

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:32]:
```suggestion
You can also only use VolumeAttributesClasses with storage backed by
{{< glossary_tooltip text="Container Storage Interface" term_id="csi" >}}, and only where the
relevant CSI driver implements the `ModifyVolume` API.
```

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:820]: (nit)
```suggestion
- `VolumeAttributesClass`: Enable support for VolumeAttributesClasses.
  See [Volume Attributes Classes](/docs/concepts/storage/volume-attributes-classes/)
  for more information.
```

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:120]:
```suggestion
VolumeAttributeClasses have parameters that describe volumes belonging to them. Different parameters may be accepted
```
_or_
Write “Volume classes” if this isn't about VolumeAttributeClasses.

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:36]: (nit)
```suggestion
Each VolumeAttributesClass contains the `driverName` and `parameters`, which are
used when a PersistentVolume (PV) belonging to the class needs to be dynamically provisioned
or modified.
```

File: content/en/docs/concepts/storage/volume-attributes-classes.md
Comment (sftim) [inline content/en/docs/concepts/storage/volume-attributes-classes.md:73]: (nit)
```suggestion
For example, a existing PersistentVolumeClaim is using a VolumeAttributesClass named silver:
```

**Review by sftim — COMMENTED** [review]
Some more feedback

**Comment by sftim** [comment]
/approve


This PR already exceeds our docs quality expectations for alpha features.

**Comment by drewhagen** [comment]
Great work today!! As a frequent reader of the Kubernetes docs, thank you for your quality contribution 😄

**Pull request merged**
